algorithms/src/Kokkos_Sort.hpp

```diff
@@ -580,48 +585,71 @@ struct min_max_functor {
+      if ((max_val - min_val) < 10000000) {
+        max_bins     = max_val - min_val + 1;
+        sort_in_bins = false;
+      }
+    }
+    if (std::is_floating_point<
+            typename ViewType::non_const_value_type>::value) {
+      KOKKOS_ASSERT(std::isfinite(static_cast<double>(result.max_val) -
+                                  static_cast<double>(result.min_val)));
+    }
+
+    BinSort<ViewType, CompType> bin_sort(
+        view, CompType(max_bins, result.min_val, result.max_val), sort_in_bins);
+    bin_sort.create_permute_vector(exec);
+    bin_sort.sort(exec, view);
+  }
+};
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <>
+class Sort<Kokkos::Cuda> {
+ public:
+  template <class ViewType>
+  static void sort(const Kokkos::Cuda&, ViewType const& view) {
+    thrust::device_ptr<typename ViewType::value_type> data_ptr(view.data());
+    thrust::sort(data_ptr, data_ptr + view.extent(0));
+  }
+};
+#endif
+
 template <class ExecutionSpace, class ViewType>
 std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value> sort(
     const ExecutionSpace& exec, ViewType const& view,
     bool const always_use_kokkos_sort = false) {
   if (!always_use_kokkos_sort) {
     if (Impl::try_std_sort(view, exec)) return;
   }
-  using CompType = BinOp1D<ViewType>;
 
-  Kokkos::MinMaxScalar<typename ViewType::non_const_value_type> result;
-  Kokkos::MinMax<typename ViewType::non_const_value_type> reducer(result);
-  parallel_reduce("Kokkos::Sort::FindExtent",
-                  Kokkos::RangePolicy<typename ViewType::execution_space>(
-                      exec, 0, view.extent(0)),
-                  Impl::min_max_functor<ViewType>(view), reducer);
-  if (result.min_val == result.max_val) return;
-  // For integral types the number of bins may be larger than the range
-  // in which case we can exactly have one unique value per bin
-  // and then don't need to sort bins.
-  bool sort_in_bins = true;
-  // TODO: figure out better max_bins then this ...
-  int64_t max_bins = view.extent(0) / 2;
-  if (std::is_integral<typename ViewType::non_const_value_type>::value) {
-    // Cast to double to avoid possible overflow when using integer
-    auto const max_val = static_cast<double>(result.max_val);
-    auto const min_val = static_cast<double>(result.min_val);
-    // using 10M as the cutoff for special behavior (roughly 40MB for the count
-    // array)
-    if ((max_val - min_val) < 10000000) {
-      max_bins     = max_val - min_val + 1;
-      sort_in_bins = false;
-    }
-  }
-  if (std::is_floating_point<typename ViewType::non_const_value_type>::value) {
-    KOKKOS_ASSERT(std::isfinite(static_cast<double>(result.max_val) -
-                                static_cast<double>(result.min_val)));
-  }
-
-  BinSort<ViewType, CompType> bin_sort(
-      view, CompType(max_bins, result.min_val, result.max_val), sort_in_bins);
-  bin_sort.create_permute_vector(exec);
-  bin_sort.sort(exec, view);
+  Sort<ExecutionSpace>::sort(exec, view);
```

Comment:
There is no point on changing the original function, you can define something like
```C++
template <class ViewType>
void sort(Cuda const& space, ViewType const& view) {
  static_assert(ViewType::rank == 1, "some useful error msg");
  auto const n = view.extent(0);
  using ValueType = typename ViewType::value_type;
  auto first = thrust::device_ptr<ValueType>(view.data());
  auto last = thrust::device_ptr<ValueType>(view.data() + n);
  auto const exec = thrust::cuda::par.on(space.cuda_stream());
  thrust::sort(exec, first, last);
}
```
which will be a better match when used with Cuda.

Reply:
is that always a better match or can there be compilers that maybe give an ambiguous template error?

Reply:
Done.